File: GoogleCast/BasicCastPlayer/objc/BasicCastPlayer/PlayerViewController.m
//
//  PlayerViewController.m
//  BasicCastPlayer
//
//  Copyright © 2019 Brightcove, Inc. All rights reserved.
//

#import "PlayerViewController.h"

@import GoogleCast;
@import BrightcovePlayerSDK;
@import BrightcoveGoogleCast;

static NSString * const kServicePolicyKey = @"BCpkADawqM0T8lW3nMChuAbrcunBBHmh4YkNl5e6ZrKQwPiK_Y83RAOF4DP5tyBF_ONBVgrEjqW6fbV0nKRuHvjRU3E8jdT9WMTOXfJODoPML6NUDCYTwTHxtNlr5YdyGYaCPLhMUZ3Xu61L";
static NSString * const kAccountID = @"5434391461001";
static NSString * const kPlaylistRefID = @"brightcove-native-sdk-plist";

@interface PlayerViewController ()<UITableViewDelegate, UITableViewDataSource, BCOVGoogleCastManagerDelegate, BCOVPlaybackControllerDelegate>

@property (nonatomic, weak) IBOutlet UITableView *tableView;
@property (nonatomic, weak) IBOutlet UIView *videoContainer;

@property (nonatomic, strong) id<BCOVPlaybackController> playbackController;
@property (nonatomic, strong) BCOVPUIPlayerView *playerView;
@property (nonatomic, strong) BCOVPlaylist *playlist;

@end

@implementation PlayerViewController

#pragma mark - View Lifecyle

- (void)viewDidLoad
{
    [super viewDidLoad];
    
    self.videoContainer.hidden = YES;
    
    GCKUICastButton *castButton = [[GCKUICastButton alloc] initWithFrame:CGRectMake(0, 0, 24, 24)];
    self.navigationItem.rightBarButtonItem = [[UIBarButtonItem alloc] initWithCustomView:castButton];

    [BCOVGoogleCastManager sharedManager].delegate = self;
    
    [self setupPlaybackController];
    [self setupPlayerView];
    [self requestPlaylist];
    
    [[NSNotificationCenter defaultCenter] addObserver:self
                                             selector:@selector(castDeviceDidChange:)
                                                 name:kGCKCastStateDidChangeNotification
                                               object:[GCKCastContext sharedInstance]];
}

#pragma mark - Misc

- (void)castDeviceDidChange:(NSNotification *)notification
{
    switch ([GCKCastContext sharedInstance].castState) {
        case GCKCastStateNoDevicesAvailable:
            NSLog(@"Cast Status: No Devices Available");
            break;
        case GCKCastStateNotConnected:
            NSLog(@"Cast Status: Not Connected");
            break;
        case GCKCastStateConnecting:
            NSLog(@"Cast Status: Connecting");
            break;
        case GCKCastStateConnected:
            NSLog(@"Cast Status: Connected");
            break;
    }
}

- (void)requestPlaylist
{
    BCOVPlaybackService *playbackService = [[BCOVPlaybackService alloc] initWithAccountId:kAccountID policyKey:kServicePolicyKey];
    
    __weak typeof(self) weakSelf = self;
    [playbackService findPlaylistWithReferenceID:kPlaylistRefID parameters:nil completion:^(BCOVPlaylist *playlist, NSDictionary *jsonResponse, NSError *error) {
       
        if (playlist)
        {
            weakSelf.playlist = playlist;
            [weakSelf.tableView reloadData];
        }
        else
        {
            NSLog(@"PlayerViewController Debug - Error retrieving video playlist");
        }
        
    }];
}

- (void)setupPlaybackController
{
    self.playbackController = [BCOVPlayerSDKManager.sharedManager createPlaybackController];
    self.playbackController.autoPlay = YES;
    self.playbackController.autoAdvance = YES;
    self.playbackController.delegate = self;
    
    [self.playbackController addSessionConsumer:BCOVGoogleCastManager.sharedManager];
}

- (void)setupPlayerView
{
    BCOVPUIPlayerViewOptions *options = [BCOVPUIPlayerViewOptions new];
    options.presentingViewController = self;
    
    BCOVPUIBasicControlView *controlView = [BCOVPUIBasicControlView basicControlViewWithVODLayout];
    self.playerView = [[BCOVPUIPlayerView alloc] initWithPlaybackController:self.playbackController options:options controlsView:controlView];
    self.playerView.translatesAutoresizingMaskIntoConstraints = NO;
    
    [self.videoContainer addSubview:self.playerView];
    
    [NSLayoutConstraint activateConstraints:@[
                                              [self.playerView.topAnchor constraintEqualToAnchor:self.videoContainer.topAnchor],
                                              [self.playerView.rightAnchor constraintEqualToAnchor:self.videoContainer.rightAnchor],
                                              [self.playerView.leftAnchor constraintEqualToAnchor:self.videoContainer.leftAnchor],
                                              [self.playerView.bottomAnchor constraintEqualToAnchor:self.videoContainer.bottomAnchor],
                                              ]];
}

#pragma mark - UITableViewDataSource

- (NSInteger)numberOfSectionsInTableView:(UITableView *)tableView
{
    return self.playlist ? 2 : 0;
}

- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section
{
    return section == 0 ? 1 : self.playlist.videos.count;
}

- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath
{
    UITableViewCell *cell = [tableView dequeueReusableCellWithIdentifier:@"BasicCell" forIndexPath:indexPath];
    
    if (indexPath.section == 0)
    {
        cell.textLabel.text = @"Play All";
    }
    
    if (indexPath.section == 1)
    {
        BCOVVideo *video = self.playlist.videos[indexPath.row];
        cell.textLabel.text = video.properties[@"name"];
    }
    
    return cell;
}

- (NSString *)tableView:(UITableView *)tableView titleForHeaderInSection:(NSInteger)section
{
    return section == 0 ? self.playlist.properties[@"name"] : nil;
}

#pragma mark - UITableViewDelegate

- (void)tableView:(UITableView *)tableView didSelectRowAtIndexPath:(NSIndexPath *)indexPath
{
    [tableView deselectRowAtIndexPath:indexPath animated:YES];
    
    self.videoContainer.hidden = [GCKCastContext sharedInstance].castState == GCKCastStateConnected;
    
    if (indexPath.section == 0)
    {
        [self.playbackController setVideos:self.playlist.videos];
        return;
    }
    
    BCOVVideo *video = self.playlist.videos[indexPath.row];
    
    [self.playbackController setVideos:@[video]];
}

#pragma mark - BCOVGoogleCastManagerDelegate

- (void)switchedToRemotePlayback
{
    self.videoContainer.hidden = YES;
}

- (void)switchedToLocalPlayback:(NSTimeInterval)lastKnownStreamPosition withError:(NSError *)error
{
    if (lastKnownStreamPosition > 0)
    {
        [self.playbackController play];
    }
    self.videoContainer.hidden = NO;
    
    if (error)
    {
        NSLog(@"Switched to local playback with error: %@", error.localizedDescription);
    }
}

- (void)currentCastedVideoDidComplete
{
    self.videoContainer.hidden = YES;
}

- (void)suitableSourceNotFound
{
    NSLog(@"Suitable source for video not found!");
}

#pragma mark - BCOVPlaybackControllerDelegate

- (void)playbackController:(id<BCOVPlaybackController>)controller playbackSession:(id<BCOVPlaybackSession>)session didReceiveLifecycleEvent:(BCOVPlaybackSessionLifecycleEvent *)lifecycleEvent
{
    if ([lifecycleEvent.eventType isEqualToString:kBCOVPlaybackSessionLifecycleEventEnd])
    {
        self.videoContainer.hidden = YES;
    }
}

@end

File: GoogleCast/BasicCastPlayer/swift/BasicCastPlayer/PlayerViewController.swift
//
//  PlayerViewController.swift
//  BasicCastPlayer
//
//  Copyright © 2019 Brightcove, Inc. All rights reserved.
//

import UIKit
import GoogleCast
import BrightcovePlayerSDK
import BrightcoveGoogleCast

fileprivate struct playbackConfig {
    static let playbackServicePolicyKey = "BCpkADawqM0T8lW3nMChuAbrcunBBHmh4YkNl5e6ZrKQwPiK_Y83RAOF4DP5tyBF_ONBVgrEjqW6fbV0nKRuHvjRU3E8jdT9WMTOXfJODoPML6NUDCYTwTHxtNlr5YdyGYaCPLhMUZ3Xu61L"
    static let accountID = "5434391461001"
    static let playlistRefID = "brightcove-native-sdk-plist"
}


@objc class PlayerViewController: UIViewController {

    @IBOutlet weak var tableView: UITableView!
    @IBOutlet weak var videoContainer: UIView!
    
    private var posters: [String:UIImage] = [:]
    private var playlist: BCOVPlaylist?
    
    lazy var playerView: BCOVPUIPlayerView? = {
        
        let options = BCOVPUIPlayerViewOptions()
        options.presentingViewController = self
        
        // Create PlayerUI views with normal VOD controls.
        let controlView = BCOVPUIBasicControlView.withVODLayout()
        guard let _playerView = BCOVPUIPlayerView(playbackController: nil, options: options, controlsView: controlView) else {
            return nil
        }
        
        // Add to parent view
        self.videoContainer.addSubview(_playerView)
        _playerView.translatesAutoresizingMaskIntoConstraints = false
        NSLayoutConstraint.activate([
            _playerView.topAnchor.constraint(equalTo: self.videoContainer.topAnchor),
            _playerView.rightAnchor.constraint(equalTo: self.videoContainer.rightAnchor),
            _playerView.leftAnchor.constraint(equalTo: self.videoContainer.leftAnchor),
            _playerView.bottomAnchor.constraint(equalTo: self.videoContainer.bottomAnchor)
            ])
        
        return _playerView
    }()
    
    lazy var playbackController: BCOVPlaybackController? = {
    
        guard let _playbackController = BCOVPlayerSDKManager.shared()?.createPlaybackController() else {
            return nil
        }
        
        _playbackController.isAutoAdvance = true
        _playbackController.isAutoPlay = true
        _playbackController.delegate = self
        
        _playbackController.add(BCOVGoogleCastManager.shared())
        
        return _playbackController
        
    }()
    
    // MARK: - View Lifecycle
    
    override func viewDidLoad() {
        super.viewDidLoad()
        
        videoContainer.isHidden = true
        
        let castButton = GCKUICastButton(frame: CGRect(x: 0, y: 0, width: 24, height: 24))
        navigationItem.rightBarButtonItem = UIBarButtonItem(customView: castButton)
        
        NotificationCenter.default.addObserver(self, selector: #selector(self.castStateDidChange),
                                               name: NSNotification.Name.gckCastStateDidChange,
                                               object: GCKCastContext.sharedInstance())
        
        requestPlaylist()
        
        playerView?.playbackController = playbackController
        
        BCOVGoogleCastManager.shared().delegate = self
    }

    // MARK: - Misc
    
    private func requestPlaylist() {
        let playbackService = BCOVPlaybackService(accountId: playbackConfig.accountID, policyKey: playbackConfig.playbackServicePolicyKey)
        playbackService?.findPlaylist(withReferenceID: playbackConfig.playlistRefID, parameters: nil, completion: { [weak self] (playlist: BCOVPlaylist?, json: [AnyHashable:Any]?, error: Error?) in
            
            guard let playlist = playlist else {
                print("PlayerViewController Debug - Error retrieving video playlist")
                return
            }
            
            self?.playlist = playlist
            self?.tableView.reloadData()
            
        })
    }
    
    // MARK: - Notification Handlers
    
    @objc private func castStateDidChange(_ notification: Notification) {
        let state = GCKCastContext.sharedInstance().castState
        
        switch state {
        case .noDevicesAvailable:
            print("No cast devices available")
        case .connected:
            print("Cast device connected")
        case .connecting:
            print("Cast device connecting")
        case .notConnected:
            print("Cast device not connected")
        }
    }

}

// MARK: - UITableViewDelegate

extension PlayerViewController: UITableViewDelegate {
    
    func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) {
        tableView.deselectRow(at: indexPath, animated: true)
        
        videoContainer.isHidden = GCKCastContext.sharedInstance().castState == .connected
        
        if let videos = playlist?.videos, indexPath.section == 0 {
            playbackController?.setVideos(videos as NSFastEnumeration)
            return
        }
        
        if let video = self.playlist?.videos[indexPath.row] as? BCOVVideo {
        
            playbackController?.setVideos([video] as NSFastEnumeration)

        }
        
    }
    
}

// MARK: - UITableViewDataSource

extension PlayerViewController: UITableViewDataSource {
    
    func numberOfSections(in tableView: UITableView) -> Int {
        if let _ = playlist {
            return 2
        }
        return 0
    }
    
    func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        if let videos = playlist?.videos {
            return section == 0 ? 1 : videos.count
        }
        return 0
    }
    
    func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
        let cell = tableView.dequeueReusableCell(withIdentifier: "BasicCell", for: indexPath)
        
        if indexPath.section == 0 {
            cell.textLabel?.text = "Play All"
            return cell
        }
        
        guard let playlist = playlist, let video = playlist.videos[indexPath.row] as? BCOVVideo, let name = video.properties["name"] as? String else {
            return cell
        }

        cell.textLabel?.text = name

        return cell
    }
    
    func tableView(_ tableView: UITableView, titleForHeaderInSection section: Int) -> String? {
        if section == 0 {
            return playlist?.properties["name"] as? String ?? nil
        }
        return nil
    }
    
}

// MARK: - BCOVPlaybackControllerDelegate

extension PlayerViewController: BCOVPlaybackControllerDelegate {
    
    func playbackController(_ controller: BCOVPlaybackController!, playbackSession session: BCOVPlaybackSession!, didReceive lifecycleEvent: BCOVPlaybackSessionLifecycleEvent!) {
        if lifecycleEvent.eventType == kBCOVPlaybackSessionLifecycleEventEnd {
            videoContainer.isHidden = true
        }
    }
    
}

// MARK: - BCOVGoogleCastManagerDelegate

extension PlayerViewController: BCOVGoogleCastManagerDelegate {
    
    func switched(toLocalPlayback lastKnownStreamPosition: TimeInterval, withError error: Error?) {
        if lastKnownStreamPosition > 0 {
            playbackController?.play()
        }
        videoContainer.isHidden = false
        
        if let _error = error {
            print("Switched to local playback with error: \(_error.localizedDescription)")
        }
    }

    func switchedToRemotePlayback() {
        videoContainer.isHidden = true
    }
    
    func currentCastedVideoDidComplete() {
        videoContainer.isHidden = true
    }
    
    func castedVideoFailedToPlay() {
        print("Failed to play casted video")
    }
    
    func suitableSourceNotFound() {
        print("Suitable source for video not found!")
    }
    
}
